packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -1389,33 +1665,268 @@ class FloatGadget : public GadgetT
     }
 };
 
-class DualVariableGadget : public libsnark::dual_variable_gadget<FieldT>
+struct SelectorGadget : public GadgetT
+{
+    const Constants& constants;
+
+    std::vector<EqualGadget> bits;
+    std::vector<UnsafeAddGadget> sum;
+
+    VariableArrayT res;
+
+    SelectorGadget(
+        ProtoboardT& pb,
+        const Constants& _constants,
+        const VariableT& type,
+        unsigned int maxBits,
+        const std::string& prefix
+    ) :
+        GadgetT(pb, prefix),
+        constants(_constants)
+    {
+        assert(maxBits < constants.values.size());
+        for (unsigned int i = 0; i < maxBits; i++)
+        {
+            bits.emplace_back(pb, type, constants.values[i], FMT(annotation_prefix, ".bits"));
+            sum.emplace_back(pb, (i == 0) ? constants._0 : sum.back().result(), bits.back().result(), FMT(annotation_prefix, ".sum"));
+            res.emplace_back(bits.back().result());
+        }
+    }
+
+    void generate_r1cs_witness()
+    {
+        for (unsigned int i = 0; i < bits.size(); i++)
+        {
+            bits[i].generate_r1cs_witness();
+            sum[i].generate_r1cs_witness();
+        }
+    }
+
+    void generate_r1cs_constraints()
+    {
+        for (unsigned int i = 0; i < bits.size(); i++)
+        {
+            bits[i].generate_r1cs_constraints();
+            sum[i].generate_r1cs_constraints();
+        }
+        // Sum needs to equal 1
+        requireEqual(pb, sum.back().result(), constants._1, FMT(annotation_prefix, ".selector_sum_one"));
+    }
+
+    const VariableArrayT& result() const
+    {
+        return res;
+    }
+};
+
+class SelectGadget : public GadgetT
 {
 public:
-    DualVariableGadget(
+
+    std::vector<TernaryGadget> results;
+
+    SelectGadget(
         ProtoboardT& pb,
-        const size_t width,
+        const VariableArrayT& selector,
+        const std::vector<VariableT>& values,
```

Comment:
> It seems we allow selectors to have all 0 values.

This gadget is only used in a single place, and the input is always the output of `SelectorGadget`which we force to have a single 1 set in the array (which is why it's called the `selector` here, perhaps the complete gadget could be used as an input here to make that more clear so it can't be misused).

> then we return [ax + by + cz] so we have 3 UnsafeAddGadget then one more constraint to check sum = cz for

That would also work. But `TernaryGadget` only uses a single constraint (if we don't need to check booleaness, which we don't want to do here in any case), so it seems the current method is easier to understand and uses 1 constraint less (not that that matters).